browser/ui/webui/new_tab_page/brave_new_tab_page_handler.cc

```diff
@@ -0,0 +1,187 @@
+  file_types.extension_description_overrides.push_back(
+      l10n_util::GetStringUTF16(IDS_UPLOAD_IMAGE_FORMAT));
+  select_file_dialog_->SelectFile(
+      ui::SelectFileDialog::SELECT_OPEN_FILE, std::u16string(),
+      profile_->last_selected_directory(), &file_types, 0,
+      base::FilePath::StringType(), web_contents_->GetTopLevelNativeWindow(),
+      nullptr);
+}
+
+void BraveNewTabPageHandler::UseBraveBackground() {
+  // Call ntp custom background images service.
+  profile_->GetPrefs()->SetBoolean(kNewTabPageCustomBackgroundEnabled, false);
+  OnCustomBackgroundImageUpdated();
+  DeleteSanitizedImageFile();
+}
+
+bool BraveNewTabPageHandler::IsCustomBackgroundEnabled() const {
+  auto* prefs = profile_->GetPrefs();
+  if (prefs->IsManagedPreference(prefs::kNtpCustomBackgroundDict))
+    return false;
+
+  return prefs->GetBoolean(kNewTabPageCustomBackgroundEnabled);
+}
+
+void BraveNewTabPageHandler::OnCustomBackgroundImageUpdated() {
+  brave_new_tab_page::mojom::CustomBackgroundPtr value =
+      brave_new_tab_page::mojom::CustomBackground::New();
+  // Pass empty struct when custom background is disabled.
+  if (IsCustomBackgroundEnabled()) {
+    // Add a timestamp to the url to prevent the browser from using a cached
+    // version when "Upload an image" is used multiple times.
+    std::string time_string = std::to_string(base::Time::Now().ToTimeT());
+    std::string local_string(ntp_background_images::kCustomWallpaperURL);
+    value->url = GURL(local_string + "?ts=" + time_string);
+  }
+  page_->OnBackgroundUpdated(std::move(value));
+}
+
+void BraveNewTabPageHandler::FileSelected(const base::FilePath& path,
+                                          int index,
+                                          void* params) {
+  profile_->set_last_selected_directory(path.DirName());
+  // By saving sanitized image, we don't need to do it whenever
+  // NTP opens.
+  ConvertSelectedImageFileAndSave(path);
+  select_file_dialog_ = nullptr;
+}
+
+void BraveNewTabPageHandler::FileSelectionCanceled(void* params) {
+  select_file_dialog_ = nullptr;
+}
+
+void BraveNewTabPageHandler::ConvertSelectedImageFileAndSave(
+    const base::FilePath& image_file) {
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(
+          [](const base::FilePath& image_file_path) {
+            std::string contents;
+            if (!base::ReadFileToString(image_file_path, &contents))
+              return absl::optional<std::string>();
+            return absl::optional<std::string>(contents);
+          },
+          image_file),
+      base::BindOnce(&BraveNewTabPageHandler::OnGotImageFile,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void BraveNewTabPageHandler::OnGotImageFile(absl::optional<std::string> input) {
+  if (!input)
+    return;
+
+  // Send image body to image decoder in isolated process.
+  GetImageDecoder()->DecodeImage(
```

Comment:
As shared in DM, this will not allow for animated GIF as it's a [SkBitmap](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/include/core/SkBitmap.h;drc=36bef2fd0b71a5f17dd5dcee9241ad705c3569e2;l=50)& 😢